Description: Use libatomic in ghc-prim
 Configure the libraries/ghc-prim/ghc-prim.cabal file to use
 libatomic if needed, same way as rts/rts.cabal does.
Author: Ilias Tsitsimpis <iliastsi@debian.org>

Index: b/libraries/ghc-prim/ghc-prim.cabal
===================================================================
--- a/libraries/ghc-prim/ghc-prim.cabal
+++ /dev/null
@@ -1,92 +0,0 @@
-cabal-version:  2.2
-name:           ghc-prim
-version:        0.9.0
--- NOTE: Don't forget to update ./changelog.md
-license:        BSD-3-Clause
-license-file:   LICENSE
-category:       GHC
-maintainer:     libraries@haskell.org
-bug-reports:    https://gitlab.haskell.org/ghc/ghc/issues/new
-synopsis:       GHC primitives
-build-type:     Custom
-description:
-    This package contains the primitive types and operations supplied by GHC.
-
-extra-source-files: changelog.md
-
-source-repository head
-    type:     git
-    location: https://gitlab.haskell.org/ghc/ghc.git
-    subdir:   libraries/ghc-prim
-
-custom-setup
-    setup-depends: base >= 4 && < 5, process, filepath, directory, Cabal >= 1.23 && < 3.9
-
-Library
-    default-language: Haskell2010
-    other-extensions:
-        BangPatterns
-        CPP
-        DeriveGeneric
-        MagicHash
-        MultiParamTypeClasses
-        NoImplicitPrelude
-        StandaloneDeriving
-        Trustworthy
-        TypeFamilies
-        UnboxedTuples
-        UnliftedFFITypes
-
-    build-depends: rts == 1.0.*
-
-    exposed-modules:
-        GHC.CString
-        GHC.Classes
-        GHC.Debug
-        GHC.Magic
-        GHC.Magic.Dict
-        GHC.Prim.Ext
-        GHC.Prim.Panic
-        GHC.Prim.Exception
-        GHC.Prim.PtrEq
-        GHC.PrimopWrappers
-        GHC.Tuple
-        GHC.Types
-
-    virtual-modules:
-        GHC.Prim
-
-    -- OS Specific
-    if os(windows)
-        -- Windows requires some extra libraries for linking because the RTS
-        -- is no longer re-exporting them (see #11223)
-        -- ucrt: standard C library. The RTS will automatically include this,
-        --       but is added for completeness.
-        -- mingw32: Unfortunately required because of a resource leak between
-        --          mingwex and mingw32. the __math_err symbol is defined in
-        --          mingw32 which is required by mingwex.
-        -- user32: provides access to apis to modify user components (UI etc)
-        --         on Windows. Required because of mingw32.
-        extra-libraries: user32, mingw32, ucrt
-
-    if os(linux)
-        -- we need libm, but for musl and other's we might need libc, as libm
-        -- is just an empty shell.
-        extra-libraries: c, m
-
-    c-sources:
-        cbits/atomic.c
-        cbits/bswap.c
-        cbits/bitrev.c
-        cbits/clz.c
-        cbits/ctz.c
-        cbits/debug.c
-        cbits/longlong.c
-        cbits/pdep.c
-        cbits/pext.c
-        cbits/popcnt.c
-        cbits/word2float.c
-
-    -- We need to set the unit ID to ghc-prim (without a version number)
-    -- as it's magic.
-    ghc-options: -this-unit-id ghc-prim
Index: b/libraries/ghc-prim/ghc-prim.cabal.in
===================================================================
--- /dev/null
+++ b/libraries/ghc-prim/ghc-prim.cabal.in
@@ -0,0 +1,98 @@
+cabal-version:  2.2
+name:           ghc-prim
+version:        0.9.0
+-- NOTE: Don't forget to update ./changelog.md
+license:        BSD-3-Clause
+license-file:   LICENSE
+category:       GHC
+maintainer:     libraries@haskell.org
+bug-reports:    https://gitlab.haskell.org/ghc/ghc/issues/new
+synopsis:       GHC primitives
+build-type:     Custom
+description:
+    This package contains the primitive types and operations supplied by GHC.
+
+extra-source-files: changelog.md
+
+source-repository head
+    type:     git
+    location: https://gitlab.haskell.org/ghc/ghc.git
+    subdir:   libraries/ghc-prim
+
+flag need-atomic
+    default: @CabalNeedLibatomic@
+
+custom-setup
+    setup-depends: base >= 4 && < 5, process, filepath, directory, Cabal >= 1.23 && < 3.9
+
+Library
+    default-language: Haskell2010
+    other-extensions:
+        BangPatterns
+        CPP
+        DeriveGeneric
+        MagicHash
+        MultiParamTypeClasses
+        NoImplicitPrelude
+        StandaloneDeriving
+        Trustworthy
+        TypeFamilies
+        UnboxedTuples
+        UnliftedFFITypes
+
+    build-depends: rts == 1.0.*
+
+    exposed-modules:
+        GHC.CString
+        GHC.Classes
+        GHC.Debug
+        GHC.Magic
+        GHC.Magic.Dict
+        GHC.Prim.Ext
+        GHC.Prim.Panic
+        GHC.Prim.Exception
+        GHC.Prim.PtrEq
+        GHC.PrimopWrappers
+        GHC.Tuple
+        GHC.Types
+
+    virtual-modules:
+        GHC.Prim
+
+    -- OS Specific
+    if os(windows)
+        -- Windows requires some extra libraries for linking because the RTS
+        -- is no longer re-exporting them (see #11223)
+        -- ucrt: standard C library. The RTS will automatically include this,
+        --       but is added for completeness.
+        -- mingw32: Unfortunately required because of a resource leak between
+        --          mingwex and mingw32. the __math_err symbol is defined in
+        --          mingw32 which is required by mingwex.
+        -- user32: provides access to apis to modify user components (UI etc)
+        --         on Windows. Required because of mingw32.
+        extra-libraries: user32, mingw32, ucrt
+
+    if os(linux)
+        -- we need libm, but for musl and other's we might need libc, as libm
+        -- is just an empty shell.
+        extra-libraries: c, m
+
+    if flag(need-atomic)
+        extra-libraries: atomic
+
+    c-sources:
+        cbits/atomic.c
+        cbits/bswap.c
+        cbits/bitrev.c
+        cbits/clz.c
+        cbits/ctz.c
+        cbits/debug.c
+        cbits/longlong.c
+        cbits/pdep.c
+        cbits/pext.c
+        cbits/popcnt.c
+        cbits/word2float.c
+
+    -- We need to set the unit ID to ghc-prim (without a version number)
+    -- as it's magic.
+    ghc-options: -this-unit-id ghc-prim
Index: b/configure.ac
===================================================================
--- a/configure.ac
+++ b/configure.ac
@@ -1236,6 +1236,7 @@ dnl When adding things to this list be s
 dnl Rules.Configure.configureResults list.
 AC_CONFIG_FILES(
 [ rts/rts.cabal
+  libraries/ghc-prim/ghc-prim.cabal
   compiler/ghc.cabal
   ghc/ghc-bin.cabal
   utils/runghc/runghc.cabal
